### PR TITLE
GROK-12941: Dendrogram: Use hamming distance in case of MSA

### PR DESCRIPTION
**Pull request by drizhina**
Use hamming distance for aligned sequences.
Refactor worker code

### CODE REVIEW

File: packages/Dendrogram/src/tests/calc-matrix-tests.ts [hunk 1]
```diff
@@ -31,12 +32,19 @@ const DNAAndNumericsDistances = [0.333, 1.118, 0, 0.333, 1, 0.833];
 const numericsDistances = [2, 4, 1, 2, 1, 3];
 const numericsDistancesTwoCols = [0.745, 1.054, 0.333, 1.054, 1, 0.667];
 
+function setMacromoleculeTags(col: DG.Column) {
+  col.semType = DG.SEMTYPE.MACROMOLECULE;
+  col.setTag(DG.TAGS.UNITS, NOTATION.FASTA);
+  col.setTag('aligned', ALIGNMENT.SEQ);
+  col.setTag('alphabet', 'DNA');
```
Comment (LaykDimon) [inline packages/Dendrogram/src/tests/calc-matrix-tests.ts:39]: consider putting tag names in constants

File: packages/Dendrogram/src/workers/distance-functions.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  }
+  return result + diff;
+}
+
+type StringDistanceFunctions = 'levenstein' | 'hamming';
+type NumberDistanceFunctions = 'numericDistance';
+
+export type DistanceFunctionNames<T> = T extends string ? StringDistanceFunctions : NumberDistanceFunctions;
+
+export const distanceFunctions: Record<DistanceFunctionNames<any>, (a:any, b:any) => number> = {
```
Comment (LaykDimon) [inline packages/Dendrogram/src/workers/distance-functions.ts:31]: no spaces before "any"

File: packages/Dendrogram/src/utils/tree-helper.ts [hunk 3]
```diff
@@ -460,16 +463,29 @@ export class TreeHelper implements ITreeHelper {
...
+    const hammingColumns: string[] = [];
     for (const col of columns) {
       let values: Float32Array;
-      if (col.type === DG.TYPE.FLOAT || col.type === DG.TYPE.INT)
-        values = await getDistanceMatrixForNumerics(col.toList() as unknown as Float32Array);
-      else if (col.semType ==='Macromolecule')
-        values = await getDistanceMatrixForSequences(col.toList() as unknown as string[]);
-      else throw new TypeError('Unsupported column type');
+      if (col.type === DG.TYPE.FLOAT || col.type === DG.TYPE.INT) {
+        values = await distanceMatrixWorker.calc(col.toList(), 'numericDistance');
```
Comment (skalkin) [inline packages/Dendrogram/src/utils/tree-helper.ts:478]: It's more efficient to use column.getRawData()

File: packages/Dendrogram/src/workers/distance-functions.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+function levensteinDistance(a: string, b: string): number {
+  return distance(a, b);
+};
+
+function numericDistance(a: number, b: number): number {
+  return Math.abs(a - b);
+}
+
+function hammingDistance(a: string, b: string) {
```
Comment (skalkin) [inline packages/Dendrogram/src/workers/distance-functions.ts:11]: All function return types should be annotated; void is the only exception from the rule

File: packages/Dendrogram/src/utils/tree-helper.ts [hunk 5]
```diff
@@ -460,22 +463,36 @@ export class TreeHelper implements ITreeHelper {
...
-        values = await getDistanceMatrixForNumerics(col.toList() as unknown as Float32Array);
-      else if (col.semType ==='Macromolecule')
-        values = await getDistanceMatrixForSequences(col.toList() as unknown as string[]);
-      else throw new TypeError('Unsupported column type');
+      if (col.type === DG.TYPE.FLOAT || col.type === DG.TYPE.INT) {
+        values = await distanceMatrixWorker.calc(col.getRawData(), 'numericDistance');
+      } else if (col.semType ==='Macromolecule') {
+        const uh = new UnitsHandler(col);
+        // Use Hamming distance when sequences are aligned
+        let seqDistanceFunction: DistanceFunctionNames<string> = 'levenstein';
```
Comment (skalkin) [inline packages/Dendrogram/src/utils/tree-helper.ts:482]: Let's use constants in such cases

File: packages/Dendrogram/src/utils/tree-helper.ts [hunk 6]
```diff
@@ -460,22 +463,36 @@ export class TreeHelper implements ITreeHelper {
...
     const columns = colNames.map((name) => df.getCol(name));
+    const hammingColumns: string[] = [];
     for (const col of columns) {
       let values: Float32Array;
-      if (col.type === DG.TYPE.FLOAT || col.type === DG.TYPE.INT)
-        values = await getDistanceMatrixForNumerics(col.toList() as unknown as Float32Array);
-      else if (col.semType ==='Macromolecule')
-        values = await getDistanceMatrixForSequences(col.toList() as unknown as string[]);
-      else throw new TypeError('Unsupported column type');
+      if (col.type === DG.TYPE.FLOAT || col.type === DG.TYPE.INT) {
```
Comment (skalkin) [inline packages/Dendrogram/src/utils/tree-helper.ts:477]: We also have DG.TYPE.QNUM; perhaps a helper method isNumerical is needed somewhere

**Pull request merged**